pallets/manta-pay/src/types.rs

```diff
@@ -184,7 +265,246 @@ impl TryFrom<SenderPost> for config::SenderPost {
+                    message: BlockArray(BoxArray(Box::new([CiphertextBlock(
+                        note.ciphertext
+                            .into_iter()
+                            .map(decode)
+                            .collect::<Result<Vec<_>, _>>()?
+                            .into(),
+                    )]))),
+                },
+            },
+        })
+    }
+}
+
+/// Incoming Note
+#[cfg_attr(
+    feature = "rpc",
+    derive(Deserialize, Serialize),
+    serde(crate = "manta_util::serde", deny_unknown_fields)
+)]
+#[derive(Clone, Debug, Decode, Default, Encode, Eq, Hash, MaxEncodedLen, PartialEq, TypeInfo)]
+pub struct LightIncomingNote {
+    /// Ephemeral Public Key
+    pub ephemeral_public_key: Group,
+
+    /// Ciphertext
+    pub ciphertext: LightIncomingCiphertext,
+}
+
+impl From<v3::LightIncomingNote> for LightIncomingNote {
+    #[inline]
+    fn from(note: v3::LightIncomingNote) -> Self {
+        let encoded = note.ciphertext.ciphertext.encode();
+        let mut encoded_arrays =
+            [[0u8; INCOMING_CIPHER_TEXT_COMPONENT_SIZE]; INCOMING_CIPHER_TEXT_COMPONENTS_COUNT];
+        for (outer_ind, array) in encoded_arrays.into_iter().enumerate() {
+            for (inner_ind, _) in array.into_iter().enumerate() {
+                encoded_arrays[outer_ind][inner_ind] =
+                    encoded[outer_ind * INCOMING_CIPHER_TEXT_COMPONENT_SIZE + inner_ind];
+            }
+        }
+        Self {
+            ephemeral_public_key: encode(note.ciphertext.ephemeral_public_key),
+            ciphertext: encoded_arrays,
+        }
+    }
+}
+
+impl TryFrom<LightIncomingNote> for v3::LightIncomingNote {
+    type Error = Error;
+
+    #[inline]
+    fn try_from(note: LightIncomingNote) -> Result<Self, Self::Error> {
+        let mut encoded_incoming_ciphertext =
+            [0u8; INCOMING_CIPHER_TEXT_COMPONENT_SIZE * INCOMING_CIPHER_TEXT_COMPONENTS_COUNT];
+        let mut ind = 0;
+        for component in note.ciphertext {
+            for byte in component {
+                encoded_incoming_ciphertext[ind as usize] = byte;
+                ind += 1;
+            }
+        }
+        let decoded_incoming_ciphertext: [u8; 96] = decode(encoded_incoming_ciphertext)?;
+        Ok(Self {
+            header: EmptyHeader::default(),
+            ciphertext: hybrid::Ciphertext {
+                ephemeral_public_key: decode(note.ephemeral_public_key)?,
+                ciphertext: decoded_incoming_ciphertext.into(),
+            },
+        })
+    }
+}
+
+/// Full Incoming Note
+#[cfg_attr(
+    feature = "rpc",
+    derive(Deserialize, Serialize),
+    serde(crate = "manta_util::serde", deny_unknown_fields)
+)]
+#[derive(Clone, Debug, Decode, Default, Encode, Eq, Hash, MaxEncodedLen, PartialEq, TypeInfo)]
+pub struct FullIncomingNote {
+    /// Address Partition
+    pub address_partition: u8,
+
+    /// Incoming Note
+    pub incoming_note: IncomingNote,
+
+    pub light_incoming_note: LightIncomingNote,
```

Comment:
why there's 2 such similar field `incoming_note` and `light_incoming_note`? The only structual diff i saw is a `tag` field, so what's the difference here?  maybe cc @bhgomes to answer.

Reply:
`IncomingNote` is computed in-circuit, and as such the soundness of its contents is guaranteed by the ZKP. This is useful if we need to audit an account e.g. imagine that someone sends you a UTXO but encrypts a corrupted `LightIncomingNote`. The UTXO is yours to spend but it won't show in your wallet, effectively hiding your assets. So someone who suspects you may request your viewing key (or may request that you open the actual `IncomingNote` with your viewing key in zero-knowledge) to check whether you control the asset in a certain UTXO (or in every UTXO in the ledger if they want).
The `LightIncomingNote` on the other hand is much faster to open because it is AES so it makes wallet sync for honest users much faster.
Both are posted on the ledger because each one has its function.
The `tag` field appears as a separate field in `IncomingNote` because of the way our duplexer construction works. The `LightIncomingNote` also has a tag coming from AES encryption, but it is part of the ciphertext.